snapcraft/_store.py

```diff
@@ -291,23 +293,69 @@ class StoreClientCLI(storeapi.StoreClient):
     #      during upload into this class using click.
     # TODO use an instance of this class directly from snapcraft.cli.store
 
-    def __init__(self, *args, **kwargs):
-        super().__init__(*args, **kwargs)
-
-        # Dynamically decorate methods that require to be logged in.
-        # This decorated logic can be replaced by proper decorated method
-        # definitions as the methods in this module are implemented within
-        # this class.
-        for method_name in [
-            "close_channels",
-            "get_account_information",
-            "push_metadata",
-            "push_precheck",
-            "register",
-            "release",
-            "upload",
-        ]:
-            setattr(self, method_name, _login_wrapper(self, getattr(self, method_name)))
+    @_login_wrapper
```

Comment:
+1i, I like this better than the setattr magic above, it's more explicit and feels more idiomatic (plus good place to add the type annotations).